bin/oneoff/data_fix/fix_invalid_users

```diff
@@ -0,0 +1,94 @@
+#!/usr/bin/env ruby
+
+require_relative '../../../dashboard/config/environment'
+
+# This script fixes invalid rows in the user table that were added to the database
+# as a result of a bug in https://github.com/code-dot-org/code-dot-org/pull/38026.
+# Briefly, there was a bug in that PR that caused user accounts created using the
+# join section endpoint (FollowersController#student_register) to be incomplete
+# or invalid (i.e. does not pass rails User model validation). The bug was present
+# only in v2020-12-02.0 which was in production from about 12/2/20 2:00pm PST to
+# 12/3/20 noon PST.
+#
+# The rows added to the user table were invalid in one of two ways:
+# 1. If the user passed valid data to the join section endpoint, the inserted row
+#    was valid except that the username field was missing. We fix these by
+#    calling generate_username to generate the username.
```

Comment:
just to double-check: we verified that a missing username is a consequence of an account being created this way, but have we verified that it is indeed the _only_ consequence?

Reply:
Hmmm, good question.  I now see in the User model that we have a bunch of validation rules that only run `on: :create` which would not have run on this group of users.  Could you take a look at the rules and tell me how important it is that we re-validate these?  (Maybe we can jump on a quick call if this is complicated.)

Reply:
Looks like 89 users are missing birthday.

Reply:
it's not the validation rules we should be looking at here; it's the various hooks that can execute. For example, usernames are generated in this `before_validation` hook: https://github.com/code-dot-org/code-dot-org/blob/10a6d27bad972fac4d3546a5993488aa770e15bb/dashboard/app/models/user.rb#L440 and the multi-auth migration takes place in this `after_create_commit` hook: https://github.com/code-dot-org/code-dot-org/blob/10a6d27bad972fac4d3546a5993488aa770e15bb/dashboard/app/models/user.rb#L132

Other similar hooks I see include: https://github.com/code-dot-org/code-dot-org/blob/10a6d27bad972fac4d3546a5993488aa770e15bb/dashboard/app/models/user.rb#L209 https://github.com/code-dot-org/code-dot-org/blob/10a6d27bad972fac4d3546a5993488aa770e15bb/dashboard/app/models/user.rb#L414-L418 https://github.com/code-dot-org/code-dot-org/blob/10a6d27bad972fac4d3546a5993488aa770e15bb/dashboard/app/models/user.rb#L452 https://github.com/code-dot-org/code-dot-org/blob/10a6d27bad972fac4d3546a5993488aa770e15bb/dashboard/app/models/user.rb#L487-L496 https://github.com/code-dot-org/code-dot-org/blob/10a6d27bad972fac4d3546a5993488aa770e15bb/dashboard/app/models/user.rb#L197 https://github.com/code-dot-org/code-dot-org/blob/10a6d27bad972fac4d3546a5993488aa770e15bb/dashboard/app/models/user.rb#L212-L216

Reply:
I should add: these are all the `before_` or `after_` hooks I was able to see in the user model, but I'm not at all convinced that this represents the complete set of things which are supposed to happen to user accounts but which did not happen correctly to these accounts. Other modules may very well add some, such as any of these: https://github.com/code-dot-org/code-dot-org/blob/10a6d27bad972fac4d3546a5993488aa770e15bb/dashboard/app/models/user.rb#L81-L86 and other parts of the creation process could add some, too

Reply:
For users that were created in this code path, `save` is called (at least) twice, the first time with `validate: false`, the second time with no additional modifiers.  This group -- the group of users that we're trying to keep -- successfully passed the second `save` call.  This means that only hooks that this group of users did not execute are the ones related to validation that are tagged with `on: :create`.  :generate_username is one of these as are a few other validation rules.